src/theory/partition_generator.cpp

```diff
@@ -53,6 +58,104 @@ PartitionGenerator::PartitionGenerator(Env& env,
   }
 }
 
+void PartitionGenerator::incrementOrInsertLemmaAtom(Node& node)
+{
+  if (d_lemmaMap.count(node) == 0)
+  {
+    d_lemmaMap.insert({node, 1});
+    d_lemmaLiterals.insert(node);
+  }
+  else
+  {
+    int new_count = d_lemmaMap[node] + 1;
+    d_lemmaMap.erase(node);
+    d_lemmaMap.insert({node, new_count});
+  }
+}
+
+void PartitionGenerator::notifyLemma(TNode n,
+                                     InferenceId id,
+                                     LemmaProperty p,
+                                     const std::vector<Node>& skAsserts,
+                                     const std::vector<Node>& sks)
+{
+  if (options().parallel.partitionStrategy == options::PartitionMode::LEMMA_CUBE
+      || options().parallel.partitionStrategy
+             == options::PartitionMode::LEMMA_SCATTER)
+  {
+    std::vector<Node> toVisit;
+    toVisit.push_back(n);
+
+    for (unsigned i = 0; i < toVisit.size(); ++i)
+    {
+      Node current = toVisit[i];
+      // If current node is theory bool, visit the children.
+      if (Theory::theoryOf(current) == THEORY_BOOL)
+      {
+        for (unsigned child = 0, childEnd = current.getNumChildren();
+             child < childEnd;
+             ++child)
+        {
+          auto childNode = current[child];
+          // If we haven't seen the child, we should visit it.
+          if (d_lemmaMap.count(childNode) == 0)
+          {
+            toVisit.push_back(childNode);
+          }
+          else
+          {
+            // If we have already seen this chld node, then it is not theory
+            // bool, so we update its entry. No need to visit again.
+            incrementOrInsertLemmaAtom(childNode);
+          }
+        }
+      }
+      else
+      {
+        incrementOrInsertLemmaAtom(current);
+      }
+    }
+  }
+}
+
+// Some of this function may be redundant, but I was trying to be
+// complete.
+bool PartitionGenerator::isUnusable(Node n)
+{
+  const std::unordered_set<Kind, kind::KindHashFunction> skolemKinds = {
+      Kind::SKOLEM};
+  const std::unordered_set<Kind, kind::KindHashFunction> unusableKinds = {
+      Kind::INST_CONSTANT};
+
+  // Check if n is constant or contains unusable kinds.
+  if (n.isConst() || expr::hasSubtermKinds(unusableKinds, n))
```

Comment:
You probably don't need the second check here, the check on line 138 is sufficient.